### PR TITLE
Feature/upgrade log2j to v2

### PR DESCRIPTION
**Pull request by PenghaiZhang**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [ ] tests are included
- [x] screenshots are included showing significant UI changes
- [ ] documentation is changed or added

##### Description of change

This is the second PR for the work of consolidating logging layer. And this PR is focused on updating Log4J to the latest version.

Overall,  the changes can be categorised into 5 groups.

1. Update Log4J to the latest version and address SBT dependency conflicts.
2. Drop code that's used to implement the custom daily rolling appender and HTML layout.
3. Rewrite Log4j configuration files in YAML format.
4. Update existing code to use the the new YAML files.
5. Update JPF plugin files to drop useless Log4j relevant configurations.

Testing was done on my local instance to confirm all logs can be properly generated.

1. All the directories in folder `logs` were created.
image

2. Archived resource centre logs were stored in folders representing different days.
image

3. The file size based rolling policy was working.
image

### CODE REVIEW

**Comment by PenghaiZhang** [comment]
Failded on `OAIEndpointTest` because the dependency `Oclc Harvester2` still uses Log4j 1.2.

**Comment by PenghaiZhang** [comment]
`Oclc Harvester2` was last updated in 2009!!!

**Comment by PenghaiZhang** [comment]
There is not updates for Oclc Harvester2 at all.

The alternative may be using OAI harvest manager which has a new release for the famous Log4J bug.

But...I can't find OAI harvest manager from maven repo.

[This is the link ](https://github.com/clarin-eric/oai-harvest-manager)

**Comment by PenghaiZhang** [comment]
I am not familiar with the OAI stuff, but I think what we can do is either replacing that OAI lib with another one which is not dependent on the old Log4J or bringing Log4J back only in this project `OldTest`.

File: Installer/data/server/learningedge-config/learningedge-log4j.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#              .....
+#
+
+Configuration:
+  status: warn
+
+  appenders:
+    RollingFile:
+      - name: FILE
+        fileName: "${install.path}/logs/resource-centre/application.html"
```
Comment (edalex-ian) [inline Installer/data/server/learningedge-config/learningedge-log4j.yaml:21]: The old templates had the filenames in properties right at the top of the file to make it easy for people to tweak. Perhaps do the same with this YAML format too?

File: Installer/data/server/learningedge-config/learningedge-log4j.yaml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        filePattern: "${install.path}/logs/resource-centre/%d{yyyy-MM-dd}/application-%i.html"
+        immediateFlush: true
+        append: true
+        Filters:
+          ThresholdFilter:
+            level: DEBUG
+          RegexFilter:
+            regex: ".*(HHH90000022|HHH90000014).*"
```
Comment (edalex-ian) [inline Installer/data/server/learningedge-config/learningedge-log4j.yaml:29]: Please add a comment defining what this is filtering.

File: Installer/data/server/manager/all/log4j.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  status: warn
+
+  properties:
+    property:
+      - name: filePath
+        value: "../logs/equella-manager"
+      - name: fileName
+        value: "${filePath}/services.html"
+      - name: filePattern
+        value: "${filePath}/%d{yyyy-MM-dd}/services-%i.html"
```
Comment (edalex-ian) [inline Installer/data/server/manager/all/log4j.yaml:11]: Just to be sure, these don't get processed by any templating in the installer do they?

File: Source/Server/equellaserver/build.sbt [hunk 4]
```diff
@@ -116,7 +116,9 @@ libraryDependencies ++= Seq(
   "javax.xml"                 % "jaxrpc"                   % "1.1",
   "jdom"                      % "jdom"                     % "1.1",
   "com.github.equella.jpf"    % "jpf"                      % "1.0.7",
-  "log4j"                     % "log4j"                    % "1.2.17",
+  "org.apache.logging.log4j"  % "log4j"                    % "2.17.1",
+  "org.apache.logging.log4j"  % "log4j-slf4j-impl"         % "2.17.1",
+  "org.apache.logging.log4j"  % "log4j-1.2-api"            % "2.17.1",
```
Comment (edalex-ian) [inline Source/Server/equellaserver/build.sbt:121]: Hmm, shouldn't this just be:

```suggestion
  "org.apache.logging.log4j"  % "log4j-api"            % "2.17.1",
```

And, do we need the API dep? Kind of feels like it should be a transitive for log4j-slf4j-impl. :thinking:

Comment (edalex-ian) [inline Source/Server/equellaserver/build.sbt:121]: Be good to pop `2.17.1` into a `val`. Maybe place that in CommonSettings... Or maybe place the whole dependency specifier in there for easy re-use - I see you also need to specify in the installer. :thinking:

File: Source/Tools/UpgradeInstallation/build.sbt [hunk 5]
```diff
@@ -1,12 +1,15 @@
...
+  "org.apache.logging.log4j" % "log4j-slf4j-impl" % "2.17.1",
+  "org.apache.logging.log4j" % "log4j"            % "2.17.1",
   xstreamDep,
   "commons-configuration" % "commons-configuration" % "1.10",
   "commons-io"            % "commons-io"            % "2.11.0",
-  "commons-lang"          % "commons-lang"          % "2.6"
+  "commons-lang"          % "commons-lang"          % "2.6",
+  // Need these two jackson deps to allow processing log4j yaml config files.
+  "com.fasterxml.jackson.core"       % "jackson-databind"        % "2.11.3",
+  "com.fasterxml.jackson.dataformat" % "jackson-dataformat-yaml" % "2.11.3"
```
Comment (edalex-ian) [inline Source/Tools/UpgradeInstallation/build.sbt:12]: How about use `jacksonVersion`?

```suggestion
  "com.fasterxml.jackson.core"       % "jackson-databind"        % jacksonVersion,
  "com.fasterxml.jackson.dataformat" % "jackson-dataformat-yaml" % jacksonVersion
```

File: docker/learningedge-log4j.yaml [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+  appenders:
+    Console:
+      name: STDOUT
+      target: SYSTEM_OUT
+      PatternLayout:
+        Pattern: "%d{ABSOLUTE} %-5p [%c{10}] %m%n"
+      Filters:
+        RegexFilter:
+          regex: ".*(HHH90000022|HHH90000014).*"
```
Comment (edalex-ian) [inline docker/learningedge-log4j.yaml:17]: As per the other, please add comment.

File: docker/learningedge-log4j.yaml [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+          onMatch: "DENY"
+          onMismatch: "NEUTRAL"
+        ThresholdFilter:
+          level: "DEBUG"
+          onMatch: "ACCEPT"
+          onMismatch: "DENY"
+
+    RollingFile:
```
Comment (edalex-ian) [inline docker/learningedge-log4j.yaml:25]: Please add the RegexFilter for this appender as well.

**Review by edalex-ian — CHANGES_REQUESTED** [review]
Great work @PenghaiZhang . Main feedback is for the dependency declarations add some variables to keep things DRY. :wink:

File: Installer/data/server/manager/all/log4j.yaml (hunk 3 above)
Comment (PenghaiZhang) [reply Installer/data/server/manager/all/log4j.yaml:11]: No, they don't. They are processed by Log4j Lookups.

Comment (PenghaiZhang) [reply Installer/data/server/manager/all/log4j.yaml:11]: image

You can see the install leaves these as they are.

File: Source/Server/equellaserver/build.sbt (hunk 4 above)
Comment (PenghaiZhang) [reply Source/Server/equellaserver/build.sbt:121]: Actually, `log4j-1.2-api` is a bridge which allows log4j2 to be used with some old log4j code. 

For example, in `ItemServiceImpl` there is usage of `MDC` which is replaced by `Thread Context Map`. And in 
`AuthenticatedThread` there is usage of `NDC` which is replaced by `Thread Context Stack`. 

I did try to update the code to use the new stuff, but I stopped and then forgot...I will try again.

File: Source/Tools/UpgradeInstallation/build.sbt (hunk 5 above)
Comment (PenghaiZhang) [reply Source/Tools/UpgradeInstallation/build.sbt:12]: The `jacksonVersion` created in `equellaserver/build.sbt` is not available here. 

`jackson-dataformat-yaml` is only used `UpgradeInstallation` so what about create a val here ?

File: Source/Server/equellaserver/build.sbt (hunk 4 above)
Comment (PenghaiZhang) [reply Source/Server/equellaserver/build.sbt]: I have moved these to `CommonSettings`.

File: Installer/data/server/manager/all/log4j.yaml (hunk 3 above)
Comment (PenghaiZhang) [reply Installer/data/server/manager/all/log4j.yaml:11]: After lots of digging, I found `learningedge-log4j.yaml`  does get processed by the installer.

The class is `Interpreter.java`.

File: Source/Tools/UpgradeInstallation/build.sbt (hunk 5 above)
Comment (edalex-ian) [reply Source/Tools/UpgradeInstallation/build.sbt:12]: Sounds like a plan. :+1:

File: Installer/data/server/learningedge-config/learningedge-log4j.yaml (hunk 1 above)
Comment (PenghaiZhang) [reply Installer/data/server/learningedge-config/learningedge-log4j.yaml:21]: As discussed, leave this file as it's now due to the complexity of refactoring how the installer processes templates.

**Review by edalex-ian — APPROVED** [review]
GTG! Thanks @PenghaiZhang :bow:

**Pull request merged**